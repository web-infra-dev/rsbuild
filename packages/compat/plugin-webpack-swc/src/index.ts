export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
