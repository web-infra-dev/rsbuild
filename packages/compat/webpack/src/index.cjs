/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
