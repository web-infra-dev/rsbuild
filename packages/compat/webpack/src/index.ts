export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
