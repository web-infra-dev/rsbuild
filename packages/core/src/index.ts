/**
 * The methods and types exported from this file are considered as
 * the public API of @rsbuild/core.
 */
import { rspack } from '@rspack/core';
import type * as Rspack from '@rspack/core';
import * as __internalHelper from './internal';

// Core methods
export { loadEnv } from './loadEnv';
export { createRsbuild } from './createRsbuild';
export { loadConfig, defineConfig } from './config';

// Rsbuild version
export const version = RSBUILD_VERSION;

// Rspack instance
export { rspack };
export type { Rspack };

// Helpers
export { logger } from '@rsbuild/shared';
export { mergeRsbuildConfig } from './mergeConfig';

// Constants
export { PLUGIN_SWC_NAME, PLUGIN_CSS_NAME } from './constants';

// Types
export type {
  // Config Types
  RsbuildConfig,
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  SourceConfig,
  ServerConfig,
  OutputConfig,
  SecurityConfig,
  PerformanceConfig,
  ModuleFederationConfig,
  // Normalized Config Types
  NormalizedConfig,
  NormalizedDevConfig,
  NormalizedHtmlConfig,
  NormalizedToolsConfig,
  NormalizedSourceConfig,
  NormalizedServerConfig,
  NormalizedOutputConfig,
  NormalizedSecurityConfig,
  NormalizedPerformanceConfig,
  NormalizedModuleFederationConfig,
  // Plugin Types
  RsbuildPlugin,
  RsbuildPlugins,
  RsbuildPluginAPI,
} from './types';

export type {
  RsbuildMode,
  RsbuildEntry,
  RsbuildTarget,
  RsbuildContext,
  RsbuildInstance,
  CreateRsbuildOptions,
  InspectConfigOptions,
  // Subtypes of Config
  Minify,
  Polyfill,
  PrintUrls,
  PublicDir,
  Decorators,
  RspackRule,
  WatchFiles,
  CSSModules,
  CrossOrigin,
  ConsoleType,
  SplitChunks,
  BundlerChain,
  ClientConfig,
  ScriptInject,
  ChainedConfig,
  PostCSSPlugin,
  ScriptLoading,
  LegalComments,
  AliasStrategy,
  FilenameConfig,
  DistPathConfig,
  OutputStructure,
  ChainIdentifier,
  PreconnectOption,
  CSSLoaderOptions,
  ModifyChainUtils,
  StyleLoaderOptions,
  PostCSSLoaderOptions,
  ChainedConfigWithUtils,
  ModifyRspackConfigUtils,
  CSSModulesLocalsConvention,
  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnAfterStartProdServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeStartProdServerFn,
  OnBeforeCreateCompilerFn,
  OnCloseDevServerFn,
  OnDevCompileDoneFn,
  ModifyBundlerChainFn,
  ModifyRspackConfigFn,
  ModifyRsbuildConfigFn,
  TransformFn,
  TransformHandler,
} from '@rsbuild/shared';

export {
  /**
   * @private
   */
  __internalHelper,
};
