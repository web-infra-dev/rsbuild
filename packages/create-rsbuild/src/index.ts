#!/usr/bin/env node

import fs from 'node:fs';
import path from 'node:path';
import {
  cancel,
  isCancel,
  multiselect,
  note,
  outro,
  select,
  text,
} from '@clack/prompts';
import deepmerge from 'deepmerge';
import { logger } from 'rslog';

function cancelAndExit() {
  cancel('Operation cancelled.');
  process.exit(0);
}

function checkCancel<T>(value: unknown) {
  if (isCancel(value)) {
    cancelAndExit();
  }
  return value as T;
}

function formatTargetDir(targetDir: string) {
  return targetDir.trim().replace(/\/+$/g, '');
}

function pkgFromUserAgent(userAgent: string | undefined) {
  if (!userAgent) return undefined;
  const pkgSpec = userAgent.split(' ')[0];
  const pkgSpecArr = pkgSpec.split('/');
  return {
    name: pkgSpecArr[0],
    version: pkgSpecArr[1],
  };
}

function isEmptyDir(path: string) {
  const files = fs.readdirSync(path);
  return files.length === 0 || (files.length === 1 && files[0] === '.git');
}

export async function main() {
  console.log('');
  logger.greet('â—†  Create Rsbuild Project');

  const cwd = process.cwd();
  const pkgInfo = pkgFromUserAgent(process.env.npm_config_user_agent);
  const pkgManager = pkgInfo ? pkgInfo.name : 'npm';
  const packageRoot = path.resolve(__dirname, '..');
  const packageJsonPath = path.join(packageRoot, 'package.json');
  const { version } = require(packageJsonPath);

  let targetDir = checkCancel<string>(
    await text({
      message: 'Input target folder',
      placeholder: 'my-project',
      validate(value) {
        if (value.length === 0) {
          return 'Target folder is required';
        }
      },
    }),
  );

  targetDir = formatTargetDir(targetDir);
  const distFolder = path.join(cwd, targetDir);

  if (fs.existsSync(distFolder) && !isEmptyDir(distFolder)) {
    const option = checkCancel<string>(
      await select({
        message: `"${targetDir}" is not empty, please choose:`,
        options: [
          { value: 'yes', label: 'Continue and override files' },
          { value: 'no', label: 'Cancel operation' },
        ],
      }),
    );

    if (option === 'no') {
      cancelAndExit();
    }
  }

  const framework = checkCancel<string>(
    await select({
      message: 'Select framework',
      options: [
        { value: 'react', label: 'React' },
        { value: 'vue3', label: 'Vue 3' },
        { value: 'vue2', label: 'Vue 2' },
        { value: 'lit', label: 'Lit' },
        { value: 'preact', label: 'Preact' },
        { value: 'svelte', label: 'Svelte' },
        { value: 'solid', label: 'Solid' },
        { value: 'vanilla', label: 'Vanilla' },
      ],
    }),
  );

  const language = checkCancel<string>(
    await select({
      message: 'Select language',
      options: [
        { value: 'ts', label: 'TypeScript' },
        { value: 'js', label: 'JavaScript' },
      ],
    }),
  );

  const tools = checkCancel<string[]>(
    await multiselect({
      message: 'Select additional tools (use arrow keys / space bar)',
      options: [
        { value: 'prettier', label: 'Add Prettier for code formatting' },
      ],
    }),
  );

  const srcFolder = path.join(packageRoot, `template-${framework}-${language}`);
  const commonFolder = path.join(packageRoot, 'template-common');

  copyFolder(commonFolder, distFolder, version);
  copyFolder(srcFolder, distFolder, version, path.basename(targetDir));

  for (const tool of tools) {
    const toolFolder = path.join(packageRoot, `template-${tool}`);
    copyFolder(toolFolder, distFolder, version);
  }

  const nextSteps = [
    `cd ${targetDir}`,
    `${pkgManager} i`,
    `${pkgManager} run dev`,
  ];

  note(nextSteps.join('\n'), 'Next steps');

  outro('Done.');
}

function sortObjectKeys(obj: Record<string, unknown>) {
  const sortedKeys = Object.keys(obj).sort();

  const sortedObj: Record<string, unknown> = {};
  for (const key of sortedKeys) {
    sortedObj[key] = obj[key];
  }

  return sortedObj;
}

function mergePackageJson(targetPackage: string, extraPackage: string) {
  if (!fs.existsSync(targetPackage)) {
    return;
  }

  const content: Record<string, unknown> = deepmerge(
    JSON.parse(fs.readFileSync(targetPackage, 'utf-8')),
    JSON.parse(fs.readFileSync(extraPackage, 'utf-8')),
  );

  for (const key of ['scripts', 'dependencies', 'devDependencies']) {
    if (!(key in content)) {
      continue;
    }
    content[key] = sortObjectKeys(content[key] as Record<string, unknown>);
  }

  fs.writeFileSync(targetPackage, `${JSON.stringify(content, null, 2)}\n`);
}

function copyFolder(src: string, dist: string, version: string, name?: string) {
  const renameFiles: Record<string, string> = {
    gitignore: '.gitignore',
  };

  // Skip local files
  const skipFiles = ['node_modules', 'dist'];

  fs.mkdirSync(dist, { recursive: true });

  for (const file of fs.readdirSync(src)) {
    if (skipFiles.includes(file)) {
      continue;
    }

    const srcFile = path.resolve(src, file);
    const distFile = renameFiles[file]
      ? path.resolve(dist, renameFiles[file])
      : path.resolve(dist, file);
    const stat = fs.statSync(srcFile);

    if (stat.isDirectory()) {
      copyFolder(srcFile, distFile, version);
    } else if (file === 'extra-package.json') {
      const targetPackage = path.resolve(dist, 'package.json');
      mergePackageJson(targetPackage, srcFile);
    } else {
      fs.copyFileSync(srcFile, distFile);

      if (file === 'package.json') {
        updatePackageJson(distFile, version, name);
      }
    }
  }
}

const updatePackageJson = (
  pkgJsonPath: string,
  version: string,
  name?: string,
) => {
  let content = fs.readFileSync(pkgJsonPath, 'utf-8');
  content = content.replace(/workspace:\*/g, `^${version}`);
  const pkg = JSON.parse(content);

  if (name && name !== '.') {
    pkg.name = name;
  }

  fs.writeFileSync(pkgJsonPath, JSON.stringify(pkg, null, 2));
};

main();
