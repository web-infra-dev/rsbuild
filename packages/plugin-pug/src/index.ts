import { mergeChainedOptions } from '@rsbuild/shared';
import type { RsbuildPlugin } from '@rsbuild/core';
import type { Options as PugOptions } from 'pug';

export type PluginPugOptions = {
  /**
   * Used to set the compilation options for Pug.
   * @see https://pugjs.org/api/reference.html#options
   */
  pugOptions?: PugOptions;
};

export const pluginPug = (options: PluginPugOptions = {}): RsbuildPlugin => ({
  name: 'rsbuild:pug',

  async setup(api) {
    const VUE_SFC_REGEXP = /\.vue$/;
    const { compile, compileClient } = await import('pug');

    const pugOptions = mergeChainedOptions({
      defaults: {
        doctype: 'html',
        compileDebug: false,
      },
      options: options.pugOptions,
    });

    api.transform(
      { test: /\.pug$/ },
      ({ code, resourcePath, addDependency }) => {
        const options = {
          filename: resourcePath,
          ...pugOptions,
        };

        // Compile pug to HTML for Vue compiler
        if (VUE_SFC_REGEXP.test(resourcePath)) {
          const template = compile(code, options);
          const { dependencies } = template as unknown as {
            dependencies: string[];
          };

          if (dependencies) {
            dependencies.forEach(addDependency);
          }

          return template();
        }

        // Compile pug to JavaScript for html-webpack-plugin
        const templateCode = compileClient(code, options);
        return `${templateCode}; export default template;`;
      },
    );
  },
});
