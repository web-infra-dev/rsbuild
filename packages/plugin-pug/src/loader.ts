import pug from 'pug';
import type { Rspack } from '@rsbuild/shared';
import type { Options } from 'pug';

const VUE_SFC_REGEXP = /\.vue$/;

export default function (this: Rspack.LoaderContext<Options>, source: string) {
  const options = {
    filename: this.resourcePath,
    doctype: 'html',
    compileDebug: false,
    ...this.getOptions(),
  };

  // Compile pug to HTML for Vue compiler
  if (VUE_SFC_REGEXP.test(this.resourcePath)) {
    const template = pug.compile(source, options);
    const { dependencies } = template as unknown as { dependencies: string[] };

    if (dependencies) {
      dependencies.forEach(this.addDependency);
    }

    return template();
  }

  // Compile pug to JavaScript for html-webpack-plugin
  const templateCode = pug.compileClient(source, options);
  return `${templateCode}; module.exports = template;`;
}
