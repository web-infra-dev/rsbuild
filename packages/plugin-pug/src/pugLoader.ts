import pug from 'pug';
import type { Rspack } from '@rsbuild/shared';
import type { Options } from 'pug';

export default function (this: Rspack.LoaderContext<Options>, source: string) {
  const options = {
    filename: this.resourcePath,
    doctype: 'html',
    compileDebug: false,
    ...this.getOptions(),
  };

  const templateCode = pug.compileClient(source, options);
  return `${templateCode}; module.exports = template;`;
}
