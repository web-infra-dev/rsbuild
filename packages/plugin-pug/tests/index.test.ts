import { createStubRsbuild } from '@rsbuild/test-helper';
import { pluginPug } from '../src';

describe('plugin-pug', () => {
  it('should add pug loader config correctly', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginPug()],
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should allow to custom pug options', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [
        pluginPug({
          pugOptions: {
            pretty: true,
          },
        }),
      ],
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });
});
