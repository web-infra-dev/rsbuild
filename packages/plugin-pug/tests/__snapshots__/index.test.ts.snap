// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-pug > should add pug loader config correctly 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.pug\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-pug/src/pugLoader",
          },
        ],
      },
    ],
  },
}
`;

exports[`plugin-pug > should allow to custom pug options 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.pug\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-pug/src/pugLoader",
            "options": {
              "pretty": true,
            },
          },
        ],
      },
    ],
  },
}
`;
